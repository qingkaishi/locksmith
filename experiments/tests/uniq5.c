#include <pthread.h>
#include <stdio.h>

char *g;
pthread_mutex_t l;

void * dostuff(void *a) {
  char *s = malloc(3);
  char *s2;
  s[1] = 3; // unique
  pthread_mutex_lock(&l);
  g = s;    // makes s global, but prior access unique; no race
  *g = 1;
  pthread_mutex_unlock(&l);
  return NULL;
}
int main() {
  pthread_t t1;
  pthread_t t2;
  pthread_mutex_init(&l,NULL);
  pthread_create(&t1, NULL, dostuff, NULL);
  pthread_create(&t2, NULL, dostuff, NULL);
}


