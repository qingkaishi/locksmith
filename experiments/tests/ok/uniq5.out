************************* STARTING *************************

shared: &g:uniq5.c:4
        alloc:uniq5.c:8
&g:uniq5.c:4 is protected by:
  concrete l:uniq5.c:20
  l:uniq5.c:5
alloc:uniq5.c:8 is protected by:
  concrete l:uniq5.c:20
  l:uniq5.c:5
*************************** DONE ***************************
